spark-3.1-spanner-lib/src/main/java/com/google/cloud/spark/spanner/SpannerSpark.java

```diff
@@ -0,0 +1,229 @@
+        switch (typename) {
+            case "ARRAY":
+                return DataTypes.ArrayType;
+            case "BOOL":
+                return DataTypes.BooleanType;
+            case "BYTES":
+                return DataTypes.BinaryType;
+            case "DATE":
+                return DataTypes.DateType;
+            case "FLOAT64":
+                return DataTypes.DecimalType;
+            case "INT64":
+                return DataTypes.LongType;
+            case "JSON":
+                return DataTypes.BinaryType;
+            case "NUMERIC":
+                return DataTypes.DecimalType;
+            case "STRING":
+                return DataTypes.StringType;
+            case "STRUCT":
+                return DataTypes.StructType;
+            case "TIMESTAMP":
+                return DataTypes.TimestampType;
+            default:
+                return DataTypes.NullType;
+        }
+    }
+
+    @Override
+    public Transform[] inferPartitioning(CaseInsensitiveStringMap options) {
+        return null;
+    }
+
+    @Override
+    public StructType inferSchema(CaseInsensitiveStringMap options) {
+        return null;
+    }
+
+    @Override
+    public boolean supportsExternalMetadata() {
+        return false;
+    }
+
+    @Override
+    public Table getTable(StructType schema, Transform[] partitioning, java.util.Map<String, String> properties) {
+        // 1. Create the DatabaseClient with the provided options:
+        // "instanceId": <INSTANCE_ID>
+        // "projectId": <PROJECT_ID>
+        // "databaseId": <DATABASE_ID>
+        String spannerUri = String.format(
+                "cloudspanner:/projects/%s/instances/%s/databases/%s", 
+                    properties.get("projectId"),
+                    properties.get("instanceId"),
+                    properties.get("databaseId"));
+
+        ConnectionOptions.Builder builder = ConnectionOptions.newBuilder().setUri(spannerUri);
+        String gcpCredsUrl = properties.get("credentials");
+        if (gcpCredsUrl!= null) {
+            builder = builder.setCredentialsUrl(gcpCredsUrl);
+        }
+        ConnectionOptions opts = builder.build();
+
+        try (Spanner spanner = opts.getService()) {
+            // 2. Acquire the Cloud Spanner client.
+            DatabaseClient dbClient = spanner.getDatabaseClient(opts.getDatabaseId());
+
+            String tableName = properties.get("table");
+            // 3. Run an information schema query to get the type definition of the table.
+            Statement stmt = Statement.of(
+                    String.format(
+                        """
+                        SELECT
+                            COLUMN_NAME, ORDINAL_POSITION, IS_NULLABLE, SPANNER_TYPE
+                        FROM
+                            INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME="%s"
+                        ORDER BY
+                            ORDINAL_POSITION
+                        """,
+                        tableName);
```

Comment:
Please use parameters to avoid Xss attack.  Examples: https://cloud.google.com/spanner/docs/samples/spanner-query-with-parameter

Reply:
Done, thanks!